File: src/main/java/no/fint/provider/springer/model/VurderingRepository.java
package no.fint.provider.springer.model;

import lombok.extern.slf4j.Slf4j;
import no.fint.event.model.Event;
import no.fint.event.model.ResponseStatus;
import no.fint.event.model.Status;
import no.fint.model.resource.FintLinks;
import no.fint.model.resource.utdanning.vurdering.EksamensgruppeResource;
import no.fint.model.resource.utdanning.vurdering.FravarResource;
import no.fint.model.resource.utdanning.vurdering.KarakterverdiResource;
import no.fint.model.resource.utdanning.vurdering.VurderingResource;
import no.fint.model.utdanning.vurdering.VurderingActions;
import no.fint.provider.springer.storage.SpringerRepository;
import org.apache.commons.lang3.StringUtils;
import org.springframework.stereotype.Repository;

import jakarta.annotation.PostConstruct;
import java.util.EnumMap;
import java.util.Set;
import java.util.stream.Collectors;

import static no.fint.model.utdanning.vurdering.VurderingActions.*;

@Slf4j
@Repository
public class VurderingRepository extends SpringerRepository {

    private final EnumMap<VurderingActions, Class<? extends FintLinks>> actions = new EnumMap<>(VurderingActions.class);

    @PostConstruct
    public void init() {
        actions.put(GET_ALL_EKSAMENSGRUPPE, EksamensgruppeResource.class);
        actions.put(GET_ALL_FRAVAR, FravarResource.class);
        actions.put(GET_ALL_KARAKTERVERDI, KarakterverdiResource.class);
        actions.put(GET_ALL_VURDERING, VurderingResource.class);
    }

    @Override
    public void accept(Event<FintLinks> response) {
        if (!StringUtils.contains(response.getSource(), "utdanning")) {
            log.info("Skipping {} from {}", response.getAction(), response.getSource());
            return;
        }

        VurderingActions action = valueOf(response.getAction());
        if (actions.containsKey(action)) {
            query(actions.get(action), response);
        } else {
            response.setStatus(Status.ADAPTER_REJECTED);
            response.setResponseStatus(ResponseStatus.REJECTED);
            response.setStatusCode("INVALID_ACTION");
            response.setMessage("Invalid action: " + action);
        }
    }

    @Override
    public Set<String> actions() {
        return actions.keySet()
                .stream()
                .map(Enum::name)
                .collect(Collectors.toSet());
    }
}

File: src/main/kotlin/no/fint/provider/springer/model/VurderingRepository.kt
package no.fint.provider.springer.model

import jakarta.annotation.PostConstruct
import no.fint.event.model.Event
import no.fint.event.model.ResponseStatus
import no.fint.event.model.Status
import no.fint.provider.springer.storage.SpringerRepository
import no.fint.provider.springer.storage.Wrapper
import no.novari.fint.model.resource.FintLinks
import no.novari.fint.model.resource.utdanning.vurdering.EksamensgruppeResource
import no.novari.fint.model.resource.utdanning.vurdering.KarakterverdiResource
import no.novari.fint.model.utdanning.vurdering.VurderingActions
import org.apache.commons.lang3.StringUtils
import org.springframework.data.mongodb.core.MongoTemplate
import org.springframework.stereotype.Repository
import java.util.*
import java.util.stream.Collectors

@Repository
class VurderingRepository(
    mongoTemplate: MongoTemplate,
    wrapper: Wrapper
) : SpringerRepository(mongoTemplate, wrapper) {

    private val actions: EnumMap<VurderingActions, Class<out FintLinks>> = EnumMap(VurderingActions::class.java)

    @PostConstruct
    fun init() {
        actions[VurderingActions.GET_ALL_EKSAMENSGRUPPE] = EksamensgruppeResource::class.java
        actions[VurderingActions.GET_ALL_KARAKTERVERDI] = KarakterverdiResource::class.java
    }

    override fun accept(response: Event<FintLinks>) {
        if (!StringUtils.contains(response.source, "utdanning")) {
            return
        }
        val action = VurderingActions.valueOf(response.action)
        val type = actions[action]
        if (type != null) {
            query(type, response)
        } else {
            response.status = Status.ADAPTER_REJECTED
            response.responseStatus = ResponseStatus.REJECTED
            response.statusCode = "INVALID_ACTION"
            response.message = "Invalid action: $action"
        }
    }

    override fun actions(): Set<String> = actions.keys.stream()
        .map { it.name }
        .collect(Collectors.toSet())
}
